### ISSUE TITLE
Is there equivalent of torch.no_grad?

### ISSUE THREAD

**Issue by HHalva** [google/jax#1937]
Is there a switch to turn off gradient computations? Or is the best way to do this just to move arrays from jax.numpy to numpy (W_nograd = onp.array(W))  -- To explain use-case: in the EM algo I'm implementing, some matrix W contributes to the loss via M-step in estimating the posteriors, and a gradient E-step. I would not want the parameter updates of W be influenced by their contributing to the posteriors from the M-step.

**Comment by hawkinsp** [comment]
Does `jax.lax.stop_gradient` do what you want?
https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.stop_gradient.html#jax.lax.stop_gradient

**Comment by HHalva** [comment]
Thanks! I see "stop" was the key-word! I was hoping more for something that you could have just at the top of a section of code like

with jax.stop_grads():
        path_through_the_graph

but I suppose it's not too inconvenient to have a switch in the grad function, so this will do - thanks!

**Comment by lavoiems** [comment]
> Does `jax.lax.stop_gradient` do what you want?
> https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.stop_gradient.html#jax.lax.stop_gradient

This is not equivalent. When calling torch.no_grad(), you inform the program that it does not need to compute the graph, potentially saving a lot of memory and compute. In many occasions, this function allowed me to perform computations, that I would not have been able to otherwise.

As far as I know, I don't think there is anything equivalent in JAX, unfortunately.

**Comment by mattjj** [comment]
I think `lax.stop_gradient` does what you describe. It prevents any derivative-related computation downstream of the values on which you call it.

**Comment by lavoiems** [comment]
Apologize for my ignorant question. The purpose of `torch.no_grad` is not only to disable the computation of the gradient, but also to disable the computation of the symbolic graph that pytorch uses to compute the gradient itself. Does JAX also computes and uses a symbolic graph to compute its gradients or is it a different paradigm?

If JAX computes a symbolic graph, then shouldn't `stop_gradient` be called before performing any operation?

**Comment by mattjj** [comment]
JAX's autodiff is pretty similar to PyTorch's. (JAX is a direct successor project to [the original Autograd](https://github.com/hips/autograd), which popularized the name and some of these techniques.)

Yes, with `lax.stop_gradient` you can prevent any gradient-related operations from downstream operations, including tracing / graph building. So, for example, if you write this:

```python
import jax

def f(x):
  x = jax.lax.stop_gradient(x)
  return jax.numpy.sin(x)

jax.grad(f)(3.)
```

then no gradient operations related to the `jax.numpy.sin` application are performed, meaning no graph building and no evaluation of cosine.

**Comment by lavoiems** [comment]
Thank you for your explanation. It all makes sense to me now.

JAX has, again, a very elegant way to do it.

**Comment by mattjj** [comment]
No problem. I hope it works for you! If not, let us know; we're here to help, and to fix things when they aren't good enough!

**Comment by JuanFMontesinos** [comment]
> No problem. I hope it works for you! If not, let us know; we're here to help, and to fix things when they aren't good enough!

Dear @mattjj I hope it's not too late to ask further questions.
To my understanding, this is not equivalent to torch.no_grad but torch.detach.

Once I call stop_gradient, it cuts the upstream gradient but jax cannot know whether I want to start building a graph again from this new array. For example, let's say I have certain preprocessing implemented in jax. 
```
input = make_jax_array(some_data)

preprocessed_input = apply_preprocessing(input)
detached_prep_input = jax.lax.stop_gradient(preprocessed_input)
# any other jax ops
do backpropagation
```
As far as I understand, the process in this pseudo code is as follows.
1. jax computes the preprocessing and builds a graph for "apply_preprocessing" that we can denote as graph1
2. detached_preps_input is created with a new, emtpy graph (graph2).
3. jax computes the additional ops and populate graph2.
4. backpropagation is done and (if not jitted) graph 1 and graph2 are deleted.

During the whole process, graph1 had to be kept alive as jax cannot know if I'm gonna still use graph1 at some point. 

On contrary, with torch.no_grad, graph 1 is directly never created. So if graph1 is large, that is an issue in terms of memory. Am I mistaken?

**Comment by MRiabov** [comment]
Gentlemen, does turning off autograd not stop JIT? I have a huge numerical computation program, and not having `grad` would probably make it faster and cheaper.

**Comment by jakevdp** [comment]
I think there's a key misunderstanding in this thread: PyTorch, as I understand it, always computes the computation graph for gradients unless you call `no_grad`. JAX never computes the gradients unless you explicitly transform your function with `jax.grad`. JAX is opt-in, torch is opt-out: so the equivalent of `no_grad` in JAX is to just not use `jax.grad` or other gradient transformations in the first place.

Once you're within a `jax.grad` context, the way you stop the gradient computation is with `jax.stop_gradient`, but semantically this is closer to `torch.detach` than `torch.no_grad`.

I hope that clears things up.

**Comment by JuanFMontesinos** [comment]
Dear @jakevdp 
Thanks for your reply, however this poses a very obvious question. 

Indeed pytorch builds the computation graph on the fly unless you explicitly call torch.no_grad.

How can jax generate the graph after calling jax.grad in the scenario you are describing? Namely, if I run certain operations and Jax does not attempt to trace them (i.e., building a graph), how is it possible to build the graph a posteriori.

**Comment by jakevdp** [comment]
I may not entirely understand the question, but if you transform your function with `jax.grad`, you compute gradients. If you don't transform the function with `jax.grad`, you don't compute gradients.

**Comment by JuanFMontesinos** [comment]
> I may not entirely understand the question, but if you transform your function with `jax.grad`, you compute gradients. If you don't transform the function with `jax.grad`, you don't compute gradients.

Thanks for the clarification, 
I was checking Jax's basics about autodiff and now I understand your point. I think this fact vanished in my brain through layers of abstraction and frameworks. 
Clarification for others:
![image](https://github.com/user-attachments/assets/e0f0ce7e-0498-49bb-ab64-82af62444560)  
As seen in the example from [Jax's Autodiff tutorial](https://jax.readthedocs.io/en/latest/notebooks/autodiff_cookbook.html) gradients won't be computed unless the functions is decorated by the grad operator.
